### ISSUE TITLE
Modula-3 folder accesses lines outside file

### ISSUE THREAD

**Issue by nyamatongwe** [ScintillaOrg/lexilla#131]
Almost every Modula-3 example file, including the attached short example, causes `std::out_of_range` exceptions when run through the `TestLexers` application. The exceptions are seen in the `GetLevel` and `SetLevel` methods which are called with `line` arguments up to `10` on this 6-line file.

[Foldm3.txt](https://github.com/ScintillaOrg/lexilla/files/10473570/Foldm3.txt)
~~~Modula-3
MODULE Main; 
IMPORT IO;
BEGIN
  IO.Put("Hello World\n")
END Main.

~~~

The folding structure is incorrect with a fold starting on the `END` statement and terminating on the following empty last line.

I haven't analysed the code much but it appears that line ends may be being counted twice, leading to `curLine` incrementing past file end. The implementation of `GetLevel` and `SetLevel` in Scintilla ignores out-of-bounds lines but still sees incorrect folding for Modula-3 files.

Exceptions in the testing application makes it more difficult to add tests for Modula-3 including for #128.

Since the folding structure is almost always wrong, folding could be disabled for Modula-3.

**Comment by rdipardo** [comment]
With the addition of one more guarding condition, as below, the exception can be downgraded to a `"different per-line folds"` lexing error [^1]:

~~~diff
diff --git a/lexers/LexModula.cxx b/lexers/LexModula.cxx
index 6347db7e..8967ee25 100644
--- a/lexers/LexModula.cxx
+++ b/lexers/LexModula.cxx
@@ -181,6 +181,7 @@ static void FoldModulaDoc( Sci_PositionU startPos,
 						 Accessor &styler)
 {
 	Sci_Position curLine = styler.GetLine(startPos);
+	Sci_Position nextLine = styler.LineStart(curLine + 1);
 	int curLevel = SC_FOLDLEVELBASE;
 	Sci_Position endPos = startPos + length;
 	if( curLine > 0 )
@@ -304,11 +305,12 @@ static void FoldModulaDoc( Sci_PositionU startPos,
 			}
 			curLine++;
 			curLevel = nextLevel;
-			if( IsEOL( styler, curPos ) && ( curPos == endPos - 1 ) ) {
+			if ((curLine < nextLine) && IsEOL(styler, curPos) && (curPos == endPos - 1)) {
 				styler.SetLevel( curLine, ( curLevel | curLevel << 16)
 								| SC_FOLDLEVELWHITEFLAG);
 			}
 			visChars = 0;
+			nextLine = curLine + 1;
 		}
 		curPos++;
 		style = styler.StyleAt( curPos );

~~~

That's not to suggest the underlying issue has anything to do with EOL detection. At least the *ad hoc* [`IsEOL()`](https://github.com/ScintillaOrg/lexilla/blob/8bc8ee65670da1a360c94f021fd033f270357ff8/lexers/LexModula.cxx#L83) function looks sound.

I would suspect one of the many other *ad hoc* functions that increment a passed-in reference to `curPos`:
 * [`checkStatement()`](https://github.com/ScintillaOrg/lexilla/blob/8bc8ee65670da1a360c94f021fd033f270357ff8/lexers/LexModula.cxx#L92)
 * [`checkKeyIdentOper()`](https://github.com/ScintillaOrg/lexilla/blob/8bc8ee65670da1a360c94f021fd033f270357ff8/lexers/LexModula.cxx#L135)
 * [`checkEndSemicolon()`](https://github.com/ScintillaOrg/lexilla/blob/8bc8ee65670da1a360c94f021fd033f270357ff8/lexers/LexModula.cxx#L112), etc.

With a big enough bump `curPos` could prematurely meet the `curPos == endPos - 1` condition, which guards the `Accessor::SetLevel` call sites.


[^1]: So far tested on these files only: [modula-3--prelim-tests.zip](https://github.com/ScintillaOrg/lexilla/files/10474313/modula-3--prelim-tests.zip)

**Comment by nyamatongwe** [comment]
The first addition implies `nextLine` is a position rather than a line number but its compared to `curLine` which is a line number.

**Comment by rdipardo** [comment]
Sorry, the added condition to circumvent the exception is below:

~~~diff
diff --git a/lexers/LexModula.cxx b/lexers/LexModula.cxx
index 1d010bf1..89242658 100644
--- a/lexers/LexModula.cxx
+++ b/lexers/LexModula.cxx
@@ -293,7 +293,7 @@ static void FoldModulaDoc( Sci_PositionU startPos,
 			break;
 		}

-		if( IsEOL( styler, curPos ) || ( curPos == endPos - 1 ) ) {
+		if( curLine < styler.GetLine(length) && ( IsEOL( styler, curPos ) || ( curPos == endPos - 1 ) )) {
 			int efectiveLevel = curLevel | nextLevel << 16;
 			if( visChars == 0 )
 				efectiveLevel |= SC_FOLDLEVELWHITEFLAG;

~~~

The lexing error message this will induce is precisely:

    ~/Issue131.m3:1: has different folds with \n versus \r\n line ends

Instead of `checkStatement()` and similar over-incrementing `curPos` as I thought, it looks rather like `curPos` is lagging behind `endPos`, keeping the fold loop going beyond the last valid line.

    Lexing ~/Issue131.m3
    currentPos = 12, endPos = 66, line = 1
    currentPos = 23, endPos = 66, line = 2
    currentPos = 29, endPos = 66, line = 3
    currentPos = 55, endPos = 66, line = 4
    currentPos = 65, endPos = 66, line = 5
    currentPos = 12, endPos = 13, line = 1
    currentPos = 12, endPos = 24, line = 1
    currentPos = 23, endPos = 24, line = 2
    currentPos = 23, endPos = 30, line = 2
    currentPos = 29, endPos = 30, line = 3
    currentPos = 29, endPos = 56, line = 3
    currentPos = 55, endPos = 56, line = 4
    currentPos = 55, endPos = 66, line = 4
    currentPos = 65, endPos = 66, line = 5
    currentPos = 65, endPos = 66, line = 5
    currentPos = 12, endPos = 71, line = 1
    currentPos = 13, endPos = 71, line = 2
    currentPos = 24, endPos = 71, line = 3
    currentPos = 25, endPos = 71, line = 4
    currentPos = 31, endPos = 71, line = 5
    currentPos = 32, endPos = 71, line = 6
    currentPos = 58, endPos = 71, line = 7
    terminate called after throwing an instance of 'std::out_of_range'
      what():  vector::_M_range_check: __n (which is 7) >= this->size() (which is 7)
    Aborted

**Comment by rdipardo** [comment]
With some more debug messages, it seems that:

 * the document length (`endPos`) is different each time `FoldModulaDoc` is called (because of faulty per-line styles or EOL changes, perhaps)
 * the exception is thrown only when the mistaken length exceeds the actual one, e.g. 71 > 66

~~~
Lexing ~/Issue131.m3
currentPos = 12, endPos = 66, line = 0
currentPos = 23, endPos = 66, line = 1
INCREMENT (checkStatement)....
old currentPos = 24
new currentPos = 28
currentPos = 29, endPos = 66, line = 2
currentPos = 55, endPos = 66, line = 3
INCREMENT (checkStatement)....
old currentPos = 56
new currentPos = 58
INCREMENT (checkStatement)....
old currentPos = 56
new currentPos = 58
INCREMENT (checkKeyIdentOper)....
old currentPos = 56
new currentPos = 64
currentPos = 65, endPos = 66, line = 4
currentPos = 65, endPos = 66, line = 5

=== END OF DOCUMENT ===

currentPos = 12, endPos = 13, line = 1

=== END OF DOCUMENT ===

currentPos = 23, endPos = 24, line = 1
currentPos = 23, endPos = 24, line = 2

=== END OF DOCUMENT ===

INCREMENT (checkStatement)....
old currentPos = 24
new currentPos = 28
currentPos = 29, endPos = 30, line = 2
currentPos = 29, endPos = 30, line = 3

=== END OF DOCUMENT ===

INCREMENT (checkStatement)....
old currentPos = 24
new currentPos = 28
currentPos = 55, endPos = 56, line = 3
currentPos = 55, endPos = 56, line = 4

=== END OF DOCUMENT ===

INCREMENT (checkStatement)....
old currentPos = 56
new currentPos = 58
INCREMENT (checkStatement)....
old currentPos = 56
new currentPos = 58
INCREMENT (checkKeyIdentOper)....
old currentPos = 56
new currentPos = 64
currentPos = 65, endPos = 66, line = 4
currentPos = 65, endPos = 66, line = 5

=== END OF DOCUMENT ===

INCREMENT (checkStatement)....
old currentPos = 56
new currentPos = 58
INCREMENT (checkStatement)....
old currentPos = 56
new currentPos = 58
INCREMENT (checkKeyIdentOper)....
old currentPos = 56
new currentPos = 64
currentPos = 65, endPos = 66, line = 5

=== END OF DOCUMENT ===

currentPos = 12, endPos = 71, line = 0
currentPos = 13, endPos = 71, line = 1
currentPos = 24, endPos = 71, line = 2
currentPos = 25, endPos = 71, line = 3
INCREMENT (checkStatement)....
old currentPos = 26
new currentPos = 30
currentPos = 31, endPos = 71, line = 4
currentPos = 32, endPos = 71, line = 5
currentPos = 58, endPos = 71, line = 6
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 7) >= this->size() (which is 7)
Aborted
~~~

**Comment by nyamatongwe** [comment]
The test harness tries with`\n` then `\r\n` line ends which would expand a 6 line 66-byte file to 71 bytes. The test harness also tries to lex the whole document at once and then line-by-line as that provokes issues where there is some state that is needed by later lines but is not saved into line state.

**Comment by rdipardo** [comment]
Your explanation about EOL exapansion was the hint I needed: [Issue131.zip](https://github.com/ScintillaOrg/lexilla/files/10478571/Issue131.zip)

Look again at the debug output just before the exception, where the value of `currentPos` increases by no more than 1 for three iterations, e.g.

    currentPos = 12, endPos = 71, line = 0
    currentPos = 13, endPos = 71, line = 1
    . . . 
    currentPos = 31, endPos = 71, line = 4
    currentPos = 32, endPos = 71, line = 5
    . . . 

Those logging calls were placed just before a call to `SetLevel`, so the folder thinks it's hit a line-end after advancing just once &#x2014; i.e. `IsEOL` returns `true` between the `\r` and `\n` of a DOS EOL!

Amending the test case for #129 revealed another minor folding bug: the level is always decremented in the presence of `END`. A guarding check for an elevated fold level can be dealt with in another ticket.

**Comment by nyamatongwe** [comment]
> Amending the test case for https://github.com/ScintillaOrg/lexilla/issues/129 revealed another minor folding bug: the level is always decremented in the presence of END. A guarding check for an elevated fold level can be dealt with in another ticket.

I feel its better to show where the folding goes wrong by allowing it to go negative. Maybe there should be an option to display a warning symbol for negative fold levels.

END always finishes some segment but the folder isn't matching all the segment starts.

**Comment by rdipardo** [comment]
> END always finishes some segment but the folder isn't matching all the segment starts.

Perhaps the simple addition of `TYPE` as a fold header is best. Unless I'm mistaken, it delimits a class definition like `RECORD` or `CLASS` in Pascal (Modula's ancestor), and LexPascal already treats the latter as fold headers.

<img width="386" alt="pascal-folding-analogue" src="https://user-images.githubusercontent.com/59004801/214202719-2c004a5f-b084-4dd3-9f88-7a863af6cd5c.png">

**Comment by nyamatongwe** [comment]
It might improve things for some examples but I don't think its completely correct as `TYPE` starts a type section rather than an single type. `RECORD` has its own `END`. From line 31 of:
https://github.com/modula3/cm3/blob/master/examples/facedetect/src/Haar.m3

There's a grammar at
https://www.cs.purdue.edu/homes/hosking/m3/reference/syntax.html
Even better! Railway diagrams:
http://modula3.github.io/cm3/src_reports/m3poster.pdf

**Comment by nyamatongwe** [comment]
Fixed in 5.2.2.
